### PR TITLE
Reconcile namespace and singleton in tandem

### PR DESCRIPTION
**Pull request by adrianludwin**
Passes the namespace object to the `reconcile` function to allow it to
be modified at the same time as the singleton. This will enable more
sophisticated reconciliations including self-service namespace creation
and adding labels to the namespace to indicate ancestry.

Tested: existing integ tests; manual with subsequent commits.

_See https://github.com/adrianludwin/multi-tenancy/commits/self-serve for the full context of this PR_

### CODE REVIEW

File: incubator/hnc/pkg/controllers/hierarchy_controller.go [hunk 1]
```diff
@@ -81,60 +81,30 @@ func (r *HierarchyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
-	// If the hierarchy object exists but is being deleted, we won't update it when we're finished
-	// syncing (we should sync our internal data structure anyway just in case something's changed).
-	// I'm not sure if this is the right thing to do but the kubebuilder boilerplate included this
-	// case explicitly.
-	if !inst.GetDeletionTimestamp().IsZero() {
-		log.Info("Singleton is being deleted")
-		update = false
+	ns, err := r.getNamespace(ctx, nm)
+	if err != nil {
+		log.Error(err, "Couldn't read namespace")
```
Comment (rjbez17) [inline incubator/hnc/pkg/controllers/hierarchy_controller.go]: If we get here is means the singleton exists but the namespace doesn't, wouldn't we want to cleanup the singleton before returning?

Comment (adrianludwin) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: Firstly, if the singleton doesn't exist, `getSingleton` doesn't return an error - it just returns an object that hasn't been committed to the apiserver. `getNamespace` works the same way.

Secondly, the singleton is _inside_ the namespace, so K8s won't allow it to exist if the namespace doesn't.

Does that answer your question?

Comment (rjbez17) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: Sorry, I missed the `!` in this check below:

```
		if !errors.IsNotFound(err) {
			return nil, err
		}
```

**Pull request merged**